src/Compiler/Microsoft.NET.Sdk.Razor.SourceGenerators/RazorSourceGenerator.cs

```diff
@@ -226,15 +227,36 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
                 .Where(pair => pair.Right)
                 .Select((pair, _) => pair.Left);
 
+            var isAddComponentParameterAvailable = compilation
+                // Only recompute when references change.
+                .WithLambdaComparer(static (compilationA, compilationB) =>
+                {
+                    return compilationA.References.SequenceEqual(compilationB.References);
+                },
+                static compilation =>
+                {
+                    return compilation.References.GetHashCode();
+                })
+                .Select((compilation, _) =>
+                {
+                    return compilation.GetTypesByMetadataName("Microsoft.AspNetCore.Components.Rendering.RenderTreeBuilder")
+                        .Any(static (t, compilation) => t.DeclaredAccessibility == Accessibility.Public &&
+                            // Only consider types from references (needed for the lambda comparer above to work correctly).
+                            t.ContainingAssembly is { } assembly && !SymbolEqualityComparer.Default.Equals(assembly, compilation.Assembly) &&
+                            t.GetMembers("AddComponentParameter").Any(static m => m.DeclaredAccessibility == Accessibility.Public), compilation);
+                });
```

Comment:
```suggestion
            var isAddComponentParameterAvailable = context.MetadataReferencesProvider
                .Collect()
                .Select((refs, _) =>
                {
                    var compilation = CSharpCompilation.Create("components", references: refs);
                    return compilation.GetTypesByMetadataName("Microsoft.AspNetCore.Components.Rendering.RenderTreeBuilder")
                        .Any(static (t, compilation) => t.DeclaredAccessibility == Accessibility.Public &&
                            t.GetMembers("AddComponentParameter").Any(static m => m.DeclaredAccessibility == Accessibility.Public), compilation);
                });
```

Reply:
You can just use the metadata refs directly, rather than going through the compilation. You can also optionally filter the references down to just `Microsoft.AspNetCore.Components.dll` to make it even quicker, but that would break the tests, as currently written.